icicle-source/src/Icicle/Source/Checker/Constraint.hs

```diff
@@ -805,21 +798,31 @@ generateP ann scrutTy resTy resTmTop resTm resPs ((pat, alt):rest) env
 appType
  :: (Hashable n)
  => a
+ -> Exp a n
  -> Type n
- -> (Type n, Type n)
  -> GenConstraintSet a n
+ -> Type n
  -> Gen a n (Type n, GenConstraintSet a n)
-appType ann resT (expT,actT) cons = do
-  let (tmpE,posE,datE) = decomposeT $ canonT expT
-  let (tmpA,posA,datA) = decomposeT $ canonT actT
-  let (tmpR,posR,datR) = decomposeT $ canonT resT
-  let consD            = require ann (CEquals datE datA)
-
-  (tmpR', consT) <- checkTemp (purely tmpE) (purely tmpA) (purely tmpR)
-  (posR', consP) <- checkPoss (definitely posE) (definitely posA) (definitely posR)
-
-  let t = recomposeT (tmpR', posR', datR)
-  return (t, concat [cons, consD, consT, consP])
+appType ann errExp funT cons actT
+  | let (tmpF,posF,datF) = decomposeT $ canonT funT
```

Comment:
I'm not sure whether it makes sense for functions to be wrapped inside a mode. For example, wouldn't `Element (a -> b)`, a stream of functions, potentially require first-class closures at runtime?

I'd be tempted to remove the decompose here so that wrapped functions are errors, unless you have specific examples that require it

Reply:
I think `Element (a -> b)` should be fine in almost all cases currently (with no let scan).

For example, this is working:
```
:let add a b = a + b
feature injury ~> let sev = add severity ~> sum (sev 5)
```
here `sev : Element (Possibly Double -> Possibly Double)` and can inline well.

`Aggregate (a -> b)` is a bit harder.

```
feature injury ~> let ugly = newest (add 4) ~> ugly (sum severity)
```

Currently type checks, but after the inline fails function application.

The other funny edge case I have is
```
feature injury ~> seq sum 4
```
Which obviously can't inline.